File: leetcode/data-structures-and-algorithms/heaps/973-k-closest-points-to-origin.java

// https://leetcode.com/problems/k-closest-points-to-origin/description/

import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {
    public int[][] kClosest(int[][] points, int k) {

        PriorityQueue<int[]> minHeap = new PriorityQueue<>(new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return (b[0] * b[0] + b[1] * b[1]) - (a[0] * a[0] + a[1] * a[1]);
            }
        });

        for (int[] p : points) {
            minHeap.add(p);

            if (minHeap.size() > k) {
                minHeap.remove();
            }
        }

        return minHeap.toArray(new int[0][0]);
    }
}

File: leetcode/data-structures-and-algorithms/heaps/973-k-closest-points-to-origin.ts
// https://leetcode.com/problems/k-closest-points-to-origin/description/

import { PriorityQueue } from "@datastructures-js/priority-queue";

function kClosest(points: number[][], k: number): number[][] {
  // max heap
  // we could optimize a little and don't do Math.sqrt()
  // if a == b => a**2 == b**2
  const heap = new PriorityQueue<number[]>(
    (a: number[], b: number[]) => dist(b) - dist(a)
  );

  for (const p of points) {
    heap.push(p);

    while (heap.size() > k) {
      heap.pop();
    }
  }

  return heap.toArray();
}

function dist(a: number[]): number {
  return Math.sqrt(a[0] ** 2 + a[1] ** 2);
}
